src/Functions/multiIf.cpp

```diff
@@ -257,11 +320,68 @@ class FunctionMultiIf final : public FunctionIfBase
+    /// We should read source from which instruction on each row?
+    template <typename S>
+    static void calculateInserts(std::vector<Instruction> & instructions, size_t rows, PaddedPODArray<S> & inserts)
+    {
+        for (S i = static_cast<S>(instructions.size() - 1); i >= 0; --i)
+        {
+            auto & instruction = instructions[i];
+            if (instruction.condition_always_true)
+            {
+                for (size_t row_i = 0; row_i < rows; ++row_i)
+                    inserts[row_i] = i;
+            }
+            else if (!instruction.condition_is_nullable)
+            {
+                const auto & cond_data = assert_cast<const ColumnUInt8 &>(*instruction.condition).getData();
+                for (size_t row_i = 0; row_i < rows; ++row_i)
+                {
+                    size_t condition_index = instruction.condition_is_short ? instruction.condition_index++ : row_i;
+
+                    /// Equivalent to below code. But it is able to utilize SIMD instructions.
+                    /// if (cond_data[condition_index])
+                    ///     inserts[row_i] = i;
+                    inserts[row_i] += cond_data[condition_index] * (i - inserts[row_i]);
+                }
+            }
+            else
+            {
+                const ColumnNullable & condition_nullable = assert_cast<const ColumnNullable &>(*instruction.condition);
+                const ColumnUInt8 & condition_nested = assert_cast<const ColumnUInt8 &>(condition_nullable.getNestedColumn());
+                const auto & condition_nested_data = condition_nested.getData();
+                const NullMap & condition_null_map = condition_nullable.getNullMapData();
+
+                for (size_t row_i = 0; row_i < rows; ++row_i)
+                {
+                    size_t condition_index = instruction.condition_is_short ? instruction.condition_index++ : row_i;
+                    /// Equivalent to below code. But it is able to utilize SIMD instructions.
+                    /// if (!condition_null_map[condition_index] && condition_nested.getData()[condition_index])
+                    ///     inserts[row_i] = i;
+                    inserts[row_i] += (~condition_null_map[condition_index] & condition_nested_data[condition_index])
+                        * (i - inserts[row_i]);
+                }
+            }
+        }
+    }
+
+    template <typename T, typename S>
+    static void executeInstructionsColumnar(std::vector<Instruction> & instructions, size_t rows, const MutableColumnPtr & res)
+    {
+        PaddedPODArray<S> inserts(rows, static_cast<S>(instructions.size()));
+        calculateInserts(instructions, rows, inserts);
+
+        PaddedPODArray<T> & res_data = assert_cast<ColumnVector<T> &>(*res).getData();
+        for (size_t row_i = 0; row_i < rows; ++row_i)
+        {
+            auto & instruction = instructions[inserts[row_i]];
+            auto ref = instruction.source->getDataAt(row_i);
```

Comment:
Same problem here, `instruction.source` can be short and it will lead to segfault